compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -337,6 +337,9 @@ pub struct InferCtxt<'a, 'tcx> {
     /// when we enter into a higher-ranked (`for<..>`) type or trait
     /// bound.
     universe: Cell<ty::UniverseIndex>,
+
+    normalize_fn_sig_for_diagnostic:
+        Option<Lrc<dyn Fn(&InferCtxt<'_, 'tcx>, ty::PolyFnSig<'tcx>) -> ty::PolyFnSig<'tcx>>>,
```

Comment:
`Box` instead of `Lrc`?

Reply:
`InferCtxt` has a `fork` method that we need to account for. I think that's only used in places like checking signatures, so we could probably seed that with a `None`, but I don't know if we're optimizing anything by using a `Box` here?

Reply:
ah true, then keeping `Lrc` seems fine